pkg/backup/restore/restore_manager.go

```diff
@@ -437,27 +411,54 @@ func (rm *restoreManager) checkTiKVEncryption(r *v1alpha1.Restore, tc *v1alpha1.
 	return nil
 }
 
-func (rm *restoreManager) readTiFlashAndTiKVReplicasFromBackupMeta(r *v1alpha1.Restore) (int32, int32, string, error) {
+func (rm *restoreManager) checkTiFlashAndTiKVReplicasFromBackupMeta(r *v1alpha1.Restore, tc *v1alpha1.TidbCluster) error {
 	metaInfo, err := backuputil.GetVolSnapBackupMetaData(r, rm.deps.SecretLister)
 	if err != nil {
-		return 0, 0, "GetVolSnapBackupMetaData failed", err
-	}
-
-	var tiflashReplicas, tikvReplicas int32
-
-	if metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash == nil {
-		tiflashReplicas = 0
-	} else {
-		tiflashReplicas = metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash.Replicas
+		klog.Errorf("GetVolSnapBackupMetaData failed")
+		return err
 	}
 
-	if metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV == nil {
-		tikvReplicas = 0
-	} else {
-		tikvReplicas = metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV.Replicas
+	// Check mismatch of tiflash config
+	if (metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash == nil ||
+		metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash.Replicas == 0) &&
+		tc.Spec.TiFlash != nil && tc.Spec.TiFlash.Replicas > 0 {
+		klog.Errorf("tiflash is enabled in TC but disabled in backup metadata")
+		return fmt.Errorf("tiflash replica mismatched")
+	} else if (tc.Spec.TiFlash == nil || tc.Spec.TiFlash.Replicas == 0) &&
+		metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash != nil &&
+		metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash.Replicas > 0 {
+		klog.Errorf("tiflash is disabled in TC enabled in backup metadata")
+		return fmt.Errorf("tiflash replica mismatched")
+	} else if tc.Spec.TiFlash != nil && metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash != nil &&
+		tc.Spec.TiFlash.Replicas != metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash.Replicas {
+		klog.Errorf("tiflash number in TC is %d but is %d in backup metadata", tc.Spec.TiFlash.Replicas, metaInfo.KubernetesMeta.TiDBCluster.Spec.TiFlash.Replicas)
+		return fmt.Errorf("tiflash replica mismatched")
+	}
+
+	// TiKV node must be there
+	if tc.Spec.TiKV == nil || tc.Spec.TiKV.Replicas == 0 {
+		klog.Errorf("ebs snapshot restore doesn't support cluster without tikv nodes")
+		return fmt.Errorf("restore to no tikv cluster")
+	} else if metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV == nil ||
+		metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV.Replicas == 0 {
+		klog.Errorf("backup source tc has no tikv nodes")
+		return fmt.Errorf("backup source tc has no tivk nodes")
+	} else if tc.Spec.TiKV.Replicas != metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV.Replicas {
+		klog.Errorf("mismatch tikv replicas, tc has %d, while backup has %d", tc.Spec.TiKV.Replicas, metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV)
+		return fmt.Errorf("tikv replica mismatch")
+	}
+
+	// Check volume number
+	if (tc.Spec.TiKV.StorageVolumes == nil && metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV.StorageVolumes != nil) ||
```

Comment:
Only `len(tc.Spec.TiKV.StorageVolumes) != len(metaInfo.KubernetesMeta.TiDBCluster.Spec.TiKV.StorageVolumes)` condition is enough. Because `len(nil)` is valid and its value is 0.

Reply:
fixed